src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceSetDeserializer.cs

```diff
@@ -147,10 +164,142 @@ public virtual IEnumerable ReadResourceSet(ODataResourceSetWrapper resourceSet,
+
+            IEdmModel model = readContext.Model;
+                
+            if (model == null)
+            {
+                throw Error.Argument("readContext", SRResources.ModelMissingFromReadContext);
+            }
+
+            IEdmEntityType actualType = model.FindType(typeName) as IEdmEntityType;
+            if (actualType == null)
             {
-                yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                throw new ODataException(Error.Format(SRResources.ResourceTypeNotInModel, typeName));
             }
+
+            if (actualType.IsAbstract)
+            {
+                string message = Error.Format(SRResources.CannotInstantiateAbstractResourceType, deltalink.TypeAnnotation.TypeName);
+                throw new ODataException(message);
+            }
+                        
+
+            IEdmDeltaLinkBase edmDeltaLink;
+            ODataDeltaDeletedLink deletedLink = deltaLinkWrapper.DeltaLink as ODataDeltaDeletedLink;
+            if (deletedLink != null)
+            {
+                edmDeltaLink = new EdmDeltaDeletedLink(actualType);
+            }
+            else
+            {
+                edmDeltaLink = new EdmDeltaLink(actualType);
+            }
+
+            edmDeltaLink.Source = deltalink.Source;
+            edmDeltaLink.Target = deltalink.Target;
+            edmDeltaLink.Relationship = deltalink.Relationship;
+
+            return edmDeltaLink;
+        }
+
+        private static EdmDeltaDeletedEntityObject DeSerializeDeletedEntity(ODataDeletedResource deletedResource, IEdmStructuredTypeReference elementType, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(deletedResource != null, "ODataDeletedResource should not be null");
+            
+            IEdmModel model = readContext.Model;
+
+            if (model == null)
+            {
+                throw Error.Argument("readContext", SRResources.ModelMissingFromReadContext);
+            }
+
+            string typeName = deletedResource.TypeName;
+
+            IEdmEntityType actualType = model.FindType(typeName) as IEdmEntityType;
+            if (actualType == null)
+            {
+                throw new ODataException(Error.Format(SRResources.ResourceTypeNotInModel, typeName));
+            }
+            string deletedResourceId = null;
+
+            if (deletedResource.Id != null)
+            {
+                deletedResourceId = deletedResource.Id.ToString();
+            }
+            else 
+            { 
+                ODataProperty key = deletedResource.Properties.FirstOrDefault(x => x.Name == actualType.Key().First().Name);
+
+                if(key != null)
+                {
+                    deletedResourceId = key.Value.ToString();
+                }
+            }
+
+            Contract.Assert(deletedResourceId != null, "ODataDeletedResource Id should not be null");
```

Comment:
this assumes a single-part key.  What if the key is multi-part?

Also, the id isn't the key value, it's the canonical URL formed using the key value(s).

We just checked a change into ODL that should ensure that deletedResource.Id is always non-null.